### ISSUE TITLE
Add workflow to build and run tests

### ISSUE THREAD

**Issue by cjappl** [realtime-sanitizer/radsan#4]
It would be nice if every did a quick CI smoke check. Github actions is free and pretty easy to do simple things.

On this branch and the llvm-project subproject we could try to build/test every change (using the specific subtests for radsan, thanks for adding to the README). Perhaps just starting with the llvm branch would be sufficient, as that's where the code lives.

I remember setting these up for rtlog-cpp and it not being that hard, happy to look into it if we are interested.

I know before too long I'll submit something half baked if I don't have a safety net.

**Comment by davidtrevelyan** [comment]
100% agreed! The only reason this wasn't activated originally was because I (mistakenly) thought that public repository CI minutes counted towards your personal plan, but thanks to your suggestion I just double checked and realised I'm wrong. Let's set it up. I'm very happy for you to propose a workflow file, and I'm also happy to add one early next week myself - up to you!

I have a little feature inbox/roadmap that I should probably share at some point. On that roadmap is automated docker image deployment, too. It might be good to think about setting that up at the same time - I don't think it would be too hard - I've already written a release script in the `docker` folder...

**Comment by cjappl** [comment]
Cool, I would love to see the roadmap for sure! 

Sounds reasonable. I am going to work on #3 as a primary step, so feel free to take this if you have cycles. If I get done with that and this ticket isn't assigned/in progress I will hop on it.

**Comment by cjappl** [comment]
Alright @davidtrevelyan  I think I am unable to complete this at the time due to perms issues. I looked at my fork of the repository and it allows me to add Actions and Workflows, but the radsan/radsan and radsan/llvm-project do not look the same to me. This leads me to believe the person that implements this needs some permissions I don't have


Options:
1. You, or someone else in the radsan org does this work
2. You elevate me, and I can do it


Either way all good! I will move the card back into "To Do" on trello

**Comment by davidtrevelyan** [comment]
@cjappl not sure I understand this one yet. I was expecting it to work just fine if you:

1. add a workflow to your fork of `realtime-sanitizer/radsan`, and then 
2. make a PR for it, merging into `realtime-sanitiser/radsan`? 

This pattern has worked for me in the past, at least. Is that what you're trying to do as well? 

PS I assumed that this workflow probably best lives in `realtime-sanitizer/radsan` for now, rather than our `realtime-sanitizer/llvm-project` fork, which has many workflows from upstream, and they're disabled for the moment. But I'm not 100% sure. If you think differently, do let me know :)

**Comment by cjappl** [comment]
Yeah, I think you're right with it living in `radsan`! Good point.

Let me try again tomorrow morning and see what I bump into, maybe I self ejected too quickly :) Thanks for the tips!

**Comment by cjappl** [comment]
> PS I assumed that this workflow probably best lives in realtime-sanitizer/radsan for now, rather than our realtime-sanitizer/llvm-project fork, which has many workflows from upstream, and they're disabled for the moment. But I'm not 100% sure. If you think differently, do let me know :)

Thinking on this a bit. The changes that we would make that are most likely to create some breaking change will be in the fork of llvm-project. Is there a way to trigger a `radsan` github action when it's dependent submodule is updated? 

If we had the workflow living in `radsan`, we would probably want to:

1. Ensure the submodule was fixed on a commit, not a branch

Then on every change to llvm-project

1. Merge the change
2. Create a PR in `radsan` updating the commit submodule, thus triggering the action


The pros of this are that we don't mess around with the existing llvm-project workflows, thus making it more obvious what the diff is.  We keep the radsan specific changes as bunched as possible in the parent repo.

The cons are that it's more overhead per commit, and more of a development cycle to figure out what is wrong. The commit is not stopped before it is merged, but instead discovered after the fact.

1. I put up a PR to `llvm-project` that has some breaking bug we don't catch in review, no workflow is run
2. You merge to `llvm-project`
3. You create a commit to `radsan` updating the submodule pointer to the right commit
4. You discover the bug
5. You come back to me showing the test failure so I can update the code and fix it

**Comment by cjappl** [comment]
See my first attempt here:
https://github.com/cjappl/radsan/actions/runs/7145020034/job/19459820881?pr=1

For checking in a basic workflow:
https://github.com/cjappl/radsan/pull/1


In terms of pricing, to be aware:

This workflow took a little over an hour, but I'll round down to an hour. According to [this doc](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions) we get 2000 minutes on the wimpy lowest tier Ubuntu box for free, which means we get about 33 of these runs a month without paying for something. That seems like it SHOULD (?) be fine in a general sense. 

To save some builds, we may want to do these on either PRs, OR merges to main, but not both.


We should also think about my above question in[ this comment ](https://github.com/realtime-sanitizer/radsan/issues/4#issuecomment-1845786805) about how we balance the changes in llvm-project, which are likely to actually break these builds, vs PRs and merges to radsan which aren't.

It seems like we could consider paying for beefier ubuntu boxes, or even get macs, but I don't know if it's really worth it as hopefully in the nearish future we can upstream to clang and piggyback on their more expansive CI system

**Comment by cjappl** [comment]
> Q: Should we ensure the fork submodule is pointing at a commit hash rather than at a branch name?

Sounds good, I can submit a PR for that after we suss out the rest of the questions here.


> Q: Is it OK to use an hour of our minutes allowance for radsan workflows every PR?

I think we are in the clear here!! I think this is for the organization, not the person behind the organization. The table on [this page](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions) has both a section for "Github Free"  and "Github free for organizations" implying there is a difference between the two.

Worst case we can try it and turn it off, easy to enable/disable



I'll open up the PR against the main branch and we can chat more. Seems like we are in general agreement/understanding

**Comment by cjappl** [comment]
Put up the PR. I think you will likely have to approve the workflow or similar for it to run, at least the first time.

**Comment by davidtrevelyan** [comment]
> Worst case we can try it and turn it off, easy to enable/disable

Very true!

> Put up the PR.

Amazing - thanks! It looks good to me! Happy for it to be merged when you are :)

**Comment by cjappl** [comment]
@davidtrevelyan merge at will. I think we can update the submodule when we merge this batch of changes, I think we have 3 changes in flight or almost done
